src/main/java/com/lginterview/providers/CourseProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+import java.util.*;
+
+@Service
+public class CourseProviderImpl implements CourseProvider {
+
+    private final CourseRepository courseRepository;
+    private final UserRepository userRepository;
+    private final Map<RequestType, CourseFunction<CourseRequest, Long>> courseActions = new HashMap<>();
+
+    @Autowired
+    public CourseProviderImpl(CourseRepository courseRepository, UserRepository userRepository) {
+        this.courseRepository = courseRepository;
+        this.userRepository = userRepository;
+
+        this.courseActions.put(RequestType.INIT, (this::createCourse));
+        this.courseActions.put(RequestType.SAVE, (this::updateCourse));
+        this.courseActions.put(RequestType.FINISH, (this::closeCourse));
+    }
+
+    @Override
+    @Transactional
+    public Long modifyCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        if (courseRequest == null)
+            throw new IllegalArgumentException("Course Request cannot be null");
+
+        if (courseRequest.getType() == null)
+            throw new IllegalArgumentException("Course Request Type cannot be null");
+
+        if (courseActions.containsKey(courseRequest.getType()))
+            return courseActions.get(courseRequest.getType()).apply(courseRequest);
+        else
+            throw new IllegalArgumentException("Invalid Course Request type");
+    }
+
+    private Long createCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course newCourse = courseRepository.findOne(courseRequest.getCourseId());
+        if (newCourse != null)
+            return newCourse.getId();
+
+        User courseUser = userRepository.findOne(courseRequest.getUserId());
+        if (courseUser == null)
+            throw new CourseDataRetrievalException("User not found");
+
+        newCourse = new Course();
+        newCourse.setStartDate(courseRequest.getTimestamp());
+        newCourse.setCourseUser(courseUser);
+        return courseRepository.save(newCourse).getId();
+    }
+
+    private Long updateCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+
+        Course courseToUpdate = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToUpdate == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        if (courseRequest.getScore() != null) {
+            int scoreValue = courseRequest.getScore();
+            courseToUpdate.getScores().add(scoreValue);
+            courseRepository.save(courseToUpdate);
+        }
+        return courseToUpdate.getId();
+    }
+
+    private Long closeCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course courseToClose = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToClose == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        courseToClose.setEndDate(courseRequest.getTimestamp());
```

Comment:
[MS][Trivial] No validation wheather endDate is 'older' than startedDate. In case of client bug you're data will be inconsistent.

Reply:
I Agree